clientv3/balancer.go

```diff
@@ -221,14 +248,62 @@ func (b *simpleBalancer) updateNotifyLoop() {
 
 func (b *simpleBalancer) notifyAddrs() {
 	b.mu.RLock()
-	addrs := b.addrs
+	eps := make([]addrConn, 0, len(b.addrs))
+	for _, ac := range b.addrs {
+		eps = append(eps, ac)
+	}
+	multi := len(b.addrs) > 1
+	if multi { // sort by state, last failed
+		sort.Slice(eps, func(i, j int) bool {
+			return eps[i].state < eps[j].state || (eps[i].state == eps[j].state && eps[i].lastFailed.Before(eps[j].lastFailed))
+		})
+	}
+	addrs := make([]grpc.Address, 0, len(b.addrs))
+	for _, ac := range eps {
+		addrs = append(addrs, ac.addr)
+	}
 	b.mu.RUnlock()
 	select {
 	case b.notifyCh <- addrs:
 	case <-b.stopc:
 	}
 }
 
+// Up implements gRPC Balancer interface. Client first establishes
+// connections to all given endpoints. As soon as one endpoint is
+// pinned, tear down other connections via 'notifyAddrs', which
+// triggers gRPC to call 'Balancer.Up' where the address connection
+// state has become 'Ready' and the HTTP/2 client has established
+// the transport.
+//
+// On transient network failure on "pinned" endpoint:
+//   1. gRPC marks address as 'connectivity.TransientFailure'
+//   2. Call 'resetTransport' on the failed endpoint
+//   3. Send 'errNetworkIO' to 'Balancer.Up.down'
+//   4. Update address state to 'connectivity.Connecting'
+//   5. Close 'ClientTransport' and calls 'Balancer.Close()'
+//     - pinned endpoint gets reset
+//   6. Establish new 'ClientTransport'
+//     - blackholed endpoints establish transport with no problem
+//   7. Update address state to 'connectivity.Ready'
+//   8. Call 'Balancer.Up' on the 'Ready' endpoint
+//   9. repeat until success
+//
+// For single-endpoint balancer, let gRPC retry until success.
+// For multi-endpoint balancer, simply notify a set of endpoints
+// and pin whichever endpoint gRPC 'Balancer.Up' first. And close
+// other endpoints. Client could get stuck retrying blackholed
+// endpoints, because gRPC just marks network I/O errors as
+// transient retrying until success. And it takes several seconds
+// to find the healthy one in next tries. To avoid wasting
+// retries, prioritize healthy endpoints on 'notifyAddrs':
+//   1. Initially mark all endpoints as 'Ready'
+//   2. Mark as 'TransientFailure' when 'Balancer.Up.down'
```

Comment:
It's `down func(err) {}` function returned from `Up` method to gRPC, which gets called when  network error happens.